MVP for the vaadin-checkbox

limonte opened the pull request:
Fixes https://github.com/vaadin/components-team-tasks/issues/250

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/vaadin/vaadin-checkbox/1)
<!-- Reviewable:end -->

tomivirkki commented:
Reviewed 4 of 6 files at r1, 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[vaadin-checkbox.html, line 26 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlNNqhLK96GgsLAHo4h:-KlNNqhM5qfHqKDa-Zk4:bugpiaw) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/27f954f2bf4e0aab97b3c572c327097f1fcd5d95/vaadin-checkbox.html#L26)):*
> ```HTML
>       [part="label"] {
>         display: inline-block;
>         padding: 2px 3px 0 0;
> ```

Why the magic paddings? Also, non-functional styles should be in `vaadin-checkbox-default-theme` module.

---

*[vaadin-checkbox.html, line 31 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlNP3XBox7rigj9VzJV:-KlNP3XCQ5vlhkBRCrJh:b61cpso) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/27f954f2bf4e0aab97b3c572c327097f1fcd5d95/vaadin-checkbox.html#L31)):*
> ```HTML
> 
>       [part="label"][disabled] {
>         color: #ccc;
> ```

Could we use opacity: 0.5, to align with `vaadin-text-field`? Also, non-functional styles should be in `vaadin-checkbox-default-theme` module.

---

*[vaadin-checkbox.html, line 35 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlNP9xqjV6payE6326A:-KlNP9xqjV6payE6326B:b1bgvw0) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/27f954f2bf4e0aab97b3c572c327097f1fcd5d95/vaadin-checkbox.html#L35)):*
> ```HTML
> 
>       :host([focus-ring]) [part="label"] {
>         box-shadow: 0 0 2px 2px Highlight;
> ```

non-functional styles should be in `vaadin-checkbox-default-theme` module.

---

*[vaadin-checkbox.html, line 39 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlNOOSSgNY98VAjSwG0:-KlNOOSSgNY98VAjSwG1:b-spghgn) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/27f954f2bf4e0aab97b3c572c327097f1fcd5d95/vaadin-checkbox.html#L39)):*
> ```HTML
> 
>       [part="native-checkbox"] {
>         margin-top: 1px;
> ```

If these are meant for aligning the checkbox better with the label, should we instead make the host an inline-flex and use flexbox centering?

---

*[vaadin-checkbox.html, line 47 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlNPiBGfN6lfExRGaZ5:-KlNPiBGfN6lfExRGaZ6:bl7m1er) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/27f954f2bf4e0aab97b3c572c327097f1fcd5d95/vaadin-checkbox.html#L47)):*
> ```HTML
>         id="label"
>         part="label"
>         checked$="[[checked]]"
> ```

Shouldn't `checked`, `disabled` and `indeterminate` be host attributes instead of binding them to every child element individually?

---

*[vaadin-checkbox.html, line 90 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlNMR0tb1w1b57DF6Cq:-KlNMR0tb1w1b57DF6Cr:b-3plvyt) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/27f954f2bf4e0aab97b3c572c327097f1fcd5d95/vaadin-checkbox.html#L90)):*
> ```HTML
>        * Part name  | Description
>        * -----------|----------------
>        * `checkbox` | The styleable `<span>` element with a graphical check
> ```

Missing `native-checkbox` from the list

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1)*
<!-- Sent from Reviewable.io -->

platosha commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[vaadin-checkbox.html, line 58 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlNTfYNCE7illilUeAH:-KlNTfYNCE7illilUeAI:b-61c301) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/27f954f2bf4e0aab97b3c572c327097f1fcd5d95/vaadin-checkbox.html#L58)):*
> ```HTML
>         disabled$="[[disabled]]"
>         indeterminate$="[[indeterminate]]"
>         tabindex="-1">
> ```

Native checkbox should have aria role of “presentational”.

Got a double announcement from screen readers with non-Shadow-DOM-native browsers otherwise.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1)*
<!-- Sent from Reviewable.io -->

tomivirkki commented:
*[vaadin-checkbox.html, line 52 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlNWpo70wZ6KurJeHxU:-KlNWpo70wZ6KurJeHxV:b-qizt28) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/27f954f2bf4e0aab97b3c572c327097f1fcd5d95/vaadin-checkbox.html#L52)):*
> ```HTML
>     >
>       <input
>         id="native-checkbox"
> ```

Unused id

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1)*
<!-- Sent from Reviewable.io -->

platosha commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[demo/index.html, line 19 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlNZLbp29HeYbGmV5j3:-KlNZLbp29HeYbGmV5j4:bf1rroa) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/27f954f2bf4e0aab97b3c572c327097f1fcd5d95/demo/index.html#L19)):*
> ```HTML
>     <demo-navigation></demo-navigation>
> 
>     <h3>Default checkbox</h3>
> ```

Use Titlecase for the Demo Titles, Please

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1)*
<!-- Sent from Reviewable.io -->

limonte commented:
Review status: 5 of 20 files reviewed at latest revision, 9 unresolved discussions.

---

*[vaadin-checkbox.html, line 52 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlNWpo70wZ6KurJeHxU:-KlNagdUBWgoNCVR846G:b-8egg0t) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/27f954f2bf4e0aab97b3c572c327097f1fcd5d95/vaadin-checkbox.html#L52)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary><blockquote>

Unused id
</blockquote></details>

Now used :)

---

*[vaadin-checkbox.html, line 90 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlNMR0tb1w1b57DF6Cq:-KlNak6JKAGmedn0RXvj:b-896fix) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/27f954f2bf4e0aab97b3c572c327097f1fcd5d95/vaadin-checkbox.html#L90)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary><blockquote>

Missing `native-checkbox` from the list
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1)*
<!-- Sent from Reviewable.io -->

tomivirkki commented:
*[vaadin-checkbox.html, line 52 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlNWpo70wZ6KurJeHxU:-KlNazxyD6R2TKBmC20J:b-x37wdw) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/27f954f2bf4e0aab97b3c572c327097f1fcd5d95/vaadin-checkbox.html#L52)):*
<details><summary><i>Previously, limonte (Limon Monte) wrote…</i></summary><blockquote>

Now used :)
</blockquote></details>

Well now the `id="label"` is no longer used :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1)*
<!-- Sent from Reviewable.io -->

limonte commented:
Review status: 4 of 20 files reviewed at latest revision, 8 unresolved discussions.

---

*[vaadin-checkbox.html, line 26 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlNNqhLK96GgsLAHo4h:-KlO29KwcgcpX6nlpx6P:b6sgvl3) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/27f954f2bf4e0aab97b3c572c327097f1fcd5d95/vaadin-checkbox.html#L26)):*
> Why the magic paddings? 

For the right gap between label and highlight box-shadow: https://i.imgur.com/o46Q5pP.png

> Also, non-functional styles should be in vaadin-checkbox-default-theme module.
 
Moved to the `default-theme`

---

*[vaadin-checkbox.html, line 31 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlNP3XBox7rigj9VzJV:-KlO2SUpffwGVDIsCnK7:b-896fix) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/27f954f2bf4e0aab97b3c572c327097f1fcd5d95/vaadin-checkbox.html#L31)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary><blockquote>

Could we use opacity: 0.5, to align with `vaadin-text-field`? Also, non-functional styles should be in `vaadin-checkbox-default-theme` module.
</blockquote></details>

Done.

---

*[vaadin-checkbox.html, line 35 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlNP9xqjV6payE6326A:-KlO2UcDs_lXb1in9ALS:b-896fix) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/27f954f2bf4e0aab97b3c572c327097f1fcd5d95/vaadin-checkbox.html#L35)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary><blockquote>

non-functional styles should be in `vaadin-checkbox-default-theme` module.
</blockquote></details>

Done.

---

*[vaadin-checkbox.html, line 39 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlNOOSSgNY98VAjSwG0:-KlO2WQk30vmpzyLPqz_:b-ey9kbu) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/27f954f2bf4e0aab97b3c572c327097f1fcd5d95/vaadin-checkbox.html#L39)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary><blockquote>

If these are meant for aligning the checkbox better with the label, should we instead make the host an inline-flex and use flexbox centering?
</blockquote></details>

Nice suggestion, done.

---

*[vaadin-checkbox.html, line 47 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlNPiBGfN6lfExRGaZ5:-KlO2_bPEorCY21xKD46:brcbpog) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/27f954f2bf4e0aab97b3c572c327097f1fcd5d95/vaadin-checkbox.html#L47)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary><blockquote>

Shouldn't `checked`, `disabled` and `indeterminate` be host attributes instead of binding them to every child element individually?
</blockquote></details>

I removed `checked`, `disabled` and `indeterminate` from `<span part="checkbox"></span>`, but I will leave those for the native checkbox to keep the native fuctionality.

---

*[vaadin-checkbox.html, line 52 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlNWpo70wZ6KurJeHxU:-KlO2p_TBQlPmk_WU-tX:b-kekcrf) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/27f954f2bf4e0aab97b3c572c327097f1fcd5d95/vaadin-checkbox.html#L52)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary><blockquote>

Well now the `id="label"` is no longer used :)
</blockquote></details>

It was used in tests, but I removed `id="label"` also.

---

*[vaadin-checkbox.html, line 58 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlNTfYNCE7illilUeAH:-KlO2z4pbh6TRKdtFybK:b-t7sn8s) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/27f954f2bf4e0aab97b3c572c327097f1fcd5d95/vaadin-checkbox.html#L58)):*
<details><summary><i>Previously, platosha (Anton Platonov) wrote…</i></summary><blockquote>

Native checkbox should have aria role of “presentational”.

Got a double announcement from screen readers with non-Shadow-DOM-native browsers otherwise.
</blockquote></details>

Done, please recheck.

---

*[demo/index.html, line 19 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlNZLbp29HeYbGmV5j3:-KlO3FpMn8j1loLJTMQj:b-896fix) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/27f954f2bf4e0aab97b3c572c327097f1fcd5d95/demo/index.html#L19)):*
<details><summary><i>Previously, platosha (Anton Platonov) wrote…</i></summary><blockquote>

Use Titlecase for the Demo Titles, Please
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1)*
<!-- Sent from Reviewable.io -->

tomivirkki commented:
Review status: 4 of 20 files reviewed at latest revision, 4 unresolved discussions.

---

*[vaadin-checkbox.html, line 39 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlNOOSSgNY98VAjSwG0:-KlOCEVs7jCa20XvNjB9:b3xhofm) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/27f954f2bf4e0aab97b3c572c327097f1fcd5d95/vaadin-checkbox.html#L39)):*
<details><summary><i>Previously, limonte (Limon Monte) wrote…</i></summary><blockquote>

Nice suggestion, done.
</blockquote></details>

I mean that we could make the `:host` an inline-flex element. Then we wouldn't need the current "wrapper" element at all (and part="label" could be changed to an actual `<label for="nativeCheckbox">`)

---

*[vaadin-checkbox.html, line 47 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlNPiBGfN6lfExRGaZ5:-KlO9KhRVPoxuhR24c6D:bh0oqhv) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/27f954f2bf4e0aab97b3c572c327097f1fcd5d95/vaadin-checkbox.html#L47)):*
<details><summary><i>Previously, limonte (Limon Monte) wrote…</i></summary><blockquote>

I removed `checked`, `disabled` and `indeterminate` from `<span part="checkbox"></span>`, but I will leave those for the native checkbox to keep the native fuctionality.
</blockquote></details>

What I meant by host attributes is that the properties should `reflectToAttribute`. Then you can style the internal elements with `:host([disabled]) [part="label"] {`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1)*
<!-- Sent from Reviewable.io -->

limonte commented:
Review status: 4 of 26 files reviewed at latest revision, 4 unresolved discussions.

---

*[vaadin-checkbox.html, line 39 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlNOOSSgNY98VAjSwG0:-KlONAJ1gZs-cpKP-5io:b-4s8hqw) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/27f954f2bf4e0aab97b3c572c327097f1fcd5d95/vaadin-checkbox.html#L39)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary><blockquote>

I mean that we could make the `:host` an inline-flex element. Then we wouldn't need the current "wrapper" element at all (and part="label" could be changed to an actual `<label for="nativeCheckbox">`)
</blockquote></details>

Thanks, nice suggestion! Done.

---

*[vaadin-checkbox.html, line 47 at r3](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlNPiBGfN6lfExRGaZ5:-KlONIPhjLW2OSM3P5ER:b-4s8hqw) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/27f954f2bf4e0aab97b3c572c327097f1fcd5d95/vaadin-checkbox.html#L47)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary><blockquote>

What I meant by host attributes is that the properties should `reflectToAttribute`. Then you can style the internal elements with `:host([disabled]) [part="label"] {`
</blockquote></details>

Thanks, nice suggestion! Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1)*
<!-- Sent from Reviewable.io -->

tomivirkki commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 15 files at r4, 5 of 12 files at r5, 14 of 14 files at r6.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-:-KlOPy9RaM4xUaK96IL-:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tomivirkki commented:
*[vaadin-checkbox.html, line 49 at r7](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlOW9SkfPAcaD4guoOn:-KlOW9SkfPAcaD4guoOo:b-9ptw2l) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/b9e93624c8c2ff8c0cb370929988194313930b0a/vaadin-checkbox.html#L49)):*
> ```HTML
>   <template>
> 
>     <label part="wrapper">
> ```

Hmm...why was this reverted back?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1)*
<!-- Sent from Reviewable.io -->

limonte commented:
@tomivirkki Anton even without testing figured out that the label won't work correctly for polyfilled browsers (e.g. Firefox). So, the rule is: do not use `<label for="<id">`, because `id` will be the same on all checkboxes and click on label will trigger the first checkbox on a page.

---

Review status: 19 of 26 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-:-KlOVf8a9MKmXGnpJBxn:b-6lrrt7)*
<!-- Sent from Reviewable.io -->

limonte commented:
Review status: 19 of 26 files reviewed at latest revision, 3 unresolved discussions.

---

*[vaadin-checkbox.html, line 49 at r7](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlOW9SkfPAcaD4guoOn:-KlOWIcOCQWEBBzaoU8n:b60m9hc) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/b9e93624c8c2ff8c0cb370929988194313930b0a/vaadin-checkbox.html#L49)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary><blockquote>

Hmm...why was this reverted back?
</blockquote></details>

Just left the comment about this chage above ^

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1)*
<!-- Sent from Reviewable.io -->

tomivirkki commented:
Right, forgot about the id in shady thing. But still, let's align with text-field and have it like this https://github.com/vaadin/vaadin-text-field/blob/master/vaadin-text-field.html#L97

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-:-KlOWTmcq1J19Ay7UMkH:blhqggp)*
<!-- Sent from Reviewable.io -->

tomivirkki commented:
*[vaadin-checkbox.html, line 49 at r7](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlOW9SkfPAcaD4guoOn:-KlOZTgYgBXOhhG1t4kv:bhvrtag) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/b9e93624c8c2ff8c0cb370929988194313930b0a/vaadin-checkbox.html#L49)):*
<details><summary><i>Previously, limonte (Limon Monte) wrote…</i></summary><blockquote>

Just left the comment about this chage above ^
</blockquote></details>

Can you do one more revert and instead of `for="id"` use the `on-click="focus"` as in text-field :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1)*
<!-- Sent from Reviewable.io -->

manolo commented:
I feel that we could share some code from text-field and button. Probably we could have a couple of mixins: `IsInputMixin` `HasActiveMixin`  WDYT?

---

Review status: 19 of 26 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[vaadin-checkbox.html, line 128 at r7](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlSAP08_TPrv_FuAX_3:-KlSAP08_TPrv_FuAX_4:b-vh1tjm) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/b9e93624c8c2ff8c0cb370929988194313930b0a/vaadin-checkbox.html#L128)):*
> ```HTML
>              * True if the checkbox is disabled.
>              */
>             disabled: {
> ```

Already defined in the control-state mixin, pls remove it

---

*[vaadin-checkbox.html, line 143 at r7](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlSAruY8r0y7t2Wm6la:-KlSAruY8r0y7t2Wm6lb:b-f2yo9h) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/b9e93624c8c2ff8c0cb370929988194313930b0a/vaadin-checkbox.html#L143)):*
> ```HTML
> 
>             /**
>              * The value of the checkbox.
> ```

This description is not clear enough. It is used only by iron-form.
Probably we should say something like:

It's the value given to the data submitted with the checkbox's name to the server when the control is inside a form.

---

*[vaadin-checkbox.html, line 149 at r7](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlSDVKZz5HcrZ-5foc5:-KlSDVKZz5HcrZ-5foc6:bheocmn) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/b9e93624c8c2ff8c0cb370929988194313930b0a/vaadin-checkbox.html#L149)):*
> ```HTML
>               type: String,
>               value: ''
>             }
> ```

You need to add the name property here, like we have in text-field

---

*[vaadin-checkbox.html, line 165 at r7](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlSNr4ZvAnHzyQZ7yV6:-KlSNr4ZvAnHzyQZ7yV7:b-pa0011) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/b9e93624c8c2ff8c0cb370929988194313930b0a/vaadin-checkbox.html#L165)):*
> ```HTML
>         }
> 
>         _addActiveListeners() {
> ```

Why is this block different to the button ? 
If we need the preventDefault, I think it should be as well in the button.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-:-KlSOKojzODmLW_Z-huL:bdgmkzk)*
<!-- Sent from Reviewable.io -->

limonte commented:
Could you please elaborate a little bit those mixins? What parts of code will be in them?

---

Review status: 19 of 26 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[vaadin-checkbox.html, line 143 at r7](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlSAruY8r0y7t2Wm6la:-KlSZ7uAEvrUwNoAp_C4:b-67up3q) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/b9e93624c8c2ff8c0cb370929988194313930b0a/vaadin-checkbox.html#L143)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

This description is not clear enough. It is used only by iron-form.
Probably we should say something like:

It's the value given to the data submitted with the checkbox's name to the server when the control is inside a form.

</blockquote></details>

`vaadin-checkbox` can not be used in a native way in forms as the `value` isn't binded to the native checkbox, suggested by you description will mislead users.

---

*[vaadin-checkbox.html, line 165 at r7](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlSNr4ZvAnHzyQZ7yV6:-KlSYmr16SYafTh2b_uK:b-e3ohkd) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/b9e93624c8c2ff8c0cb370929988194313930b0a/vaadin-checkbox.html#L165)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

Why is this block different to the button ? 
If we need the preventDefault, I think it should be as well in the button.
</blockquote></details>

There are a couple of reasons. One of those is button can be activated by spaced or enter and checkbox only by space.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-:-KlSZa741LWxBNdPYuO5:b-r3j7i5)*
<!-- Sent from Reviewable.io -->

tomivirkki commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 19 of 26 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-:-KlSbXwX2CBzj4y4JjPk:bnfp4nl)*
<!-- Sent from Reviewable.io -->

manolo commented:
Input mixin is the one with the value and name properties as well as extending the ControlState mixin, althought  thinking a little bit more, it's so small that could be skip.
But for the active mixin is the stuff adding the active attribute that should be the same than with the button, and probably with radio button

---

Review status: 19 of 26 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[vaadin-checkbox.html, line 63 at r7](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlSbvpyE1O5BCgryKww:-KlSbvpyE1O5BCgryKwx:b9du7d8) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/b9e93624c8c2ff8c0cb370929988194313930b0a/vaadin-checkbox.html#L63)):*
> ```HTML
> 
>       <span part="label">
>         <slot></slot>
> ```

Probably we would need a helper label property to set the default text of the label ? we have label in most form components

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-:-KlSbBe2xsV4uYJbj8Tq:bk7blvz)*
<!-- Sent from Reviewable.io -->

manolo commented:
Review status: 19 of 26 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[vaadin-checkbox.html, line 143 at r7](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlSAruY8r0y7t2Wm6la:-KlShZ3cSc_Fq7e6dgG5:bmf3m7) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/b9e93624c8c2ff8c0cb370929988194313930b0a/vaadin-checkbox.html#L143)):*
<details><summary><i>Previously, limonte (Limon Monte) wrote…</i></summary><blockquote>

`vaadin-checkbox` can not be used in a native way in forms as the `value` isn't binded to the native checkbox, suggested by you description will mislead users.
</blockquote></details>

you don't follow me, I'm refering to synthetic implementations of form, like iron-form which serializes the web-component based on its name and value, it does not go inside the shadow. I don't want to bind the name to the internal native input

---

*[vaadin-checkbox.html, line 165 at r7](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlSNr4ZvAnHzyQZ7yV6:-KlSi2ZiY1dt-xy1lkMA:bla597g) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/b9e93624c8c2ff8c0cb370929988194313930b0a/vaadin-checkbox.html#L165)):*
<details><summary><i>Previously, limonte (Limon Monte) wrote…</i></summary><blockquote>

There are a couple of reasons. One of those is button can be activated by spaced or enter and checkbox only by space.
</blockquote></details>

Ok, then.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1)*
<!-- Sent from Reviewable.io -->

limonte commented:
> But for the active mixin is the stuff adding the active attribute that should be the same than with the button, and probably with radio button

As discussed before, activation could be different for different components. E.g. for the button it's space/enter and for the checkbox it's the space only. So, imo it doesn't worth to extract that functionality in the common mixin.

---

Review status: 18 of 26 files reviewed at latest revision, 8 unresolved discussions.

---

*[vaadin-checkbox.html, line 63 at r7](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlSbvpyE1O5BCgryKww:-KlSrLdcoM2HoaFLQBdg:bbf1ffm) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/b9e93624c8c2ff8c0cb370929988194313930b0a/vaadin-checkbox.html#L63)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

Probably we would need a helper label property to set the default text of the label ? we have label in most form components
</blockquote></details>

IMO there shouldn't be any default label. Users need checkbox without any labels sometimes.

---

*[vaadin-checkbox.html, line 149 at r7](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlSDVKZz5HcrZ-5foc5:-KlT-fA17MA0NDtOYbNo:b-896fix) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/b9e93624c8c2ff8c0cb370929988194313930b0a/vaadin-checkbox.html#L149)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

You need to add the name property here, like we have in text-field
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-:-KlSrf1g2gobbvaI7TsK:b-r4e9oi)*
<!-- Sent from Reviewable.io -->

tomivirkki commented:
*[vaadin-checkbox.html, line 160 at r8](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlT01pfbbGRR1INMZSH:-KlT01pfbbGRR1INMZSI:b-nfp4nq) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/c9020186d8297da8d608e7812dfaa9ff8a71c408/vaadin-checkbox.html#L160)):*
> ```HTML
>         }
> 
>         check() {
> ```

Please don't add new features/API during review phase. I already gave LGTM to this PR but it didn't include this API back then.

Are these public (undocumented) utility functions needed? IMO they bloat the API unnecessarily without providing much extra user value. Since this PR is called "MVP", should we leave this kind of stuff for later discussion.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1)*
<!-- Sent from Reviewable.io -->

limonte commented:
Review status: 18 of 26 files reviewed at latest revision, 6 unresolved discussions.

---

*[vaadin-checkbox.html, line 143 at r7](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlSAruY8r0y7t2Wm6la:-KlT3sid0K3-wLdKCKwW:b-1si0ty) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/b9e93624c8c2ff8c0cb370929988194313930b0a/vaadin-checkbox.html#L143)):*
> This description is not clear enough. It is used only by iron-form.
> Probably we should say something like:

> It's the value given to the data submitted with the checkbox's name to the server when the control is inside a form.

Done, thanks for the explanation!

---

*[vaadin-checkbox.html, line 160 at r8](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlT01pfbbGRR1INMZSH:-KlT4-f8e9ayDDteEx0C:b-93djeb) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/c9020186d8297da8d608e7812dfaa9ff8a71c408/vaadin-checkbox.html#L160)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary><blockquote>

Please don't add new features/API during review phase. I already gave LGTM to this PR but it didn't include this API back then.

Are these public (undocumented) utility functions needed? IMO they bloat the API unnecessarily without providing much extra user value. Since this PR is called "MVP", should we leave this kind of stuff for later discussion.
</blockquote></details>

Reverted, sorry.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1)*
<!-- Sent from Reviewable.io -->

limonte commented:
Review status: 18 of 26 files reviewed at latest revision, 6 unresolved discussions.

---

*[vaadin-checkbox.html, line 128 at r7](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlSAP08_TPrv_FuAX_3:-KlT44O5REju_yByk5j1:bss4a2b) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/b9e93624c8c2ff8c0cb370929988194313930b0a/vaadin-checkbox.html#L128)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

Already defined in the control-state mixin, pls remove it
</blockquote></details>

Removed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1)*
<!-- Sent from Reviewable.io -->

manolo commented:
Review status: 17 of 26 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[vaadin-checkbox.html, line 63 at r7](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlSbvpyE1O5BCgryKww:-KlTF_FEtFOFjOT03gs1:bvv8bpy) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/b9e93624c8c2ff8c0cb370929988194313930b0a/vaadin-checkbox.html#L63)):*
<details><summary><i>Previously, limonte (Limon Monte) wrote…</i></summary><blockquote>

IMO there shouldn't be any default label. Users need checkbox without any labels sometimes.
</blockquote></details>

Sure. Then they don't define label. What I meant

```
<span part="label">
     <slot>[[label]]</slot>
  </span>
```

Then the user can have:

```
<!-- default no label shown -->
<vaadin-checkbox>
<vaadin-text-field>
<!-- specify label using declarative  -->
<vaadin-checkbox label="foo"> 
<vaadin-text-field label="bar">
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1)*
<!-- Sent from Reviewable.io -->

limonte commented:
Review status: 17 of 26 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[vaadin-checkbox.html, line 54 at r7](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlSqWXi_Wu697nczNpe:-KlTH1WNg2bar8D638AD:b-dtkds3) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/b9e93624c8c2ff8c0cb370929988194313930b0a/vaadin-checkbox.html#L54)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary><blockquote>

Binding should be two-way so that checking the native checkbox affects the component's `checked` also:
`checked="{{checked::change}}"`

There should be a test that verifies this
</blockquote></details>

Done, thanks!

---

*[vaadin-checkbox.html, line 63 at r7](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlSbvpyE1O5BCgryKww:-KlTIErsU9k3ZUUpcbTd:b-4odhx6) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/b9e93624c8c2ff8c0cb370929988194313930b0a/vaadin-checkbox.html#L63)):*
> <span part="label">
>    <slot>[[label]]</slot>
> </span>

I can't make it work that way. It that way of binding a property to a slot documented somewhere? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1)*
<!-- Sent from Reviewable.io -->

manolo commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 6 files at r1, 3 of 15 files at r4, 4 of 12 files at r5, 7 of 14 files at r6, 6 of 7 files at r7, 3 of 3 files at r9.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[vaadin-checkbox.html, line 63 at r7](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlSbvpyE1O5BCgryKww:-KlTHGqUpu-Nd-wMOU3a:bubj510) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/b9e93624c8c2ff8c0cb370929988194313930b0a/vaadin-checkbox.html#L63)):*
<details><summary><i>Previously, limonte (Limon Monte) wrote…</i></summary><blockquote>

> <span part="label">
>    <slot>[[label]]</slot>
> </span>

I can't make it work that way. It that way of binding a property to a slot documented somewhere? 
</blockquote></details>

True, the unnamed slot always takes content from the innerHtml of the component.
To make it work we need `<slot name=label>` but we don't want the user to specify the slot.
So the implementation is ok for the mvp, and the label feature cannot be here without some coding.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-:-KlTM0-8ug3VmGR91l2X:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tomivirkki commented:
Reviewed 6 of 7 files at r7, 3 of 3 files at r9.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1)*
<!-- Sent from Reviewable.io -->

manolo commented:
Reviewed 1 of 1 files at r10.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1)*
<!-- Sent from Reviewable.io -->

Saulis commented:
Question about the indeterminate mode: as it behaves a bit inconsistently on different browsers, should we try to unify the behavior to be same for each browser (nice feature) or just go with the native behavior for now? I'm OK with both. 
(See https://stackoverflow.com/questions/10099158/how-to-deal-with-browser-differences-with-indeterminate-checkbox)

---

Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[vaadin-checkbox.html, line 57 at r10](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlUdMR6P35t0N5irhsp:-KlUdMR7ZEA8evjHB7NO:bmhcrbp) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/403225704c3b682d84ea7a4114a83a6bcb09f603/vaadin-checkbox.html#L57)):*
> ```HTML
>         checked="{{checked::change}}"
>         disabled$="[[disabled]]"
>         indeterminate$="[[indeterminate]]"
> ```

`indeterminate` doesn't currently work, this needs to be `indeterminate="{{indeterminate::change}}"` (need to bind to a property instead of attribute + two-way binding needed). We could also add demo and tests for the indeterminate mode.

---

*[vaadin-checkbox.html, line 119 at r10](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlUeCIGFsEpddQwf8V2:-KlUeCIGFsEpddQwf8V3:b-ao7mgo) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/403225704c3b682d84ea7a4114a83a6bcb09f603/vaadin-checkbox.html#L119)):*
> ```HTML
>              * True if the checkbox is checked.
>              */
>             checked: {
> ```

Needs `notify: true` since `checked` changes after user action

---

*[vaadin-checkbox.html, line 130 at r10](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlUeEDRPG5M01NnIyuj:-KlUeEDRPG5M01NnIyuk:b8zkg1g) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/403225704c3b682d84ea7a4114a83a6bcb09f603/vaadin-checkbox.html#L130)):*
> ```HTML
>              * https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/checkbox#Indeterminate_state_checkboxes
>              */
>             indeterminate: {
> ```

Needs `notify: true` since `indeterminate` changes after user action

---

*[vaadin-checkbox.html, line 145 at r10](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlUeOj8ts06FiY_nDgU:-KlUeOj8ts06FiY_nDgV:b-g8jovl) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/403225704c3b682d84ea7a4114a83a6bcb09f603/vaadin-checkbox.html#L145)):*
> ```HTML
>              * The value given to the data submitted with the checkbox's name to the server when the control is inside a form.
>              */
>             value: {
> ```

This isn't set or bound anywhere?

---

*[test/index.html, line 16 at r10](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlUfNYnm2KmsrmbPumT:-KlUfNYo3cNQIHoQu-kD:b840dpu) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/403225704c3b682d84ea7a4114a83a6bcb09f603/test/index.html#L16)):*
> ```HTML
>     WCT.loadSuites([
>       'vaadin-checkbox_test.html',
>       'vaadin-checkbox_test.html?wc-shadydom=true'
> ```

With Polymer 2, I'm OK running tests only in default Shadow/Shady mode per browser.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-:-KlUedilOBGc5sraO9wG:b-xbyi0p)*
<!-- Sent from Reviewable.io -->

Saulis commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[vaadin-checkbox.html, line 161 at r10](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlUh1E86oeDexgXuXIM:-KlUh1E86oeDexgXuXIN:b-8tkd6t) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/403225704c3b682d84ea7a4114a83a6bcb09f603/vaadin-checkbox.html#L161)):*
> ```HTML
> 
>         _checkedChanged(checked) {
>           this.setAttribute('aria-checked', checked);
> ```

`aria-checked` should be set to `mixed` when `indeterminate` is `true`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1)*
<!-- Sent from Reviewable.io -->

manolo commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[vaadin-checkbox.html, line 145 at r10](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlUeOj8ts06FiY_nDgU:-KlX8OFiCswBeiZoazBI:b-gcdqdo) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/403225704c3b682d84ea7a4114a83a6bcb09f603/vaadin-checkbox.html#L145)):*
<details><summary><i>Previously, Saulis (Sauli Tähkäpää) wrote…</i></summary><blockquote>

This isn't set or bound anywhere?
</blockquote></details>

It's needed in iron-form. So it's here for API doc purposes. Iron-form will ignore this if it does not have name+value. 

---

*[test/index.html, line 16 at r10](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlUfNYnm2KmsrmbPumT:-KlX8qql_O7xIeXzriOU:bsv6p8d) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/403225704c3b682d84ea7a4114a83a6bcb09f603/test/index.html#L16)):*
<details><summary><i>Previously, Saulis (Sauli Tähkäpää) wrote…</i></summary><blockquote>

With Polymer 2, I'm OK running tests only in default Shadow/Shady mode per browser.
</blockquote></details>

We are running both in other P2 elements but those that are problematic. I think if tests don't have performance problems, and in the time that shady dom is supported we should test shady.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1)*
<!-- Sent from Reviewable.io -->

limonte commented:
> Question about the indeterminate mode: as it behaves a bit inconsistently on different browsers, should we try to unify the behavior to be same for each browser (nice feature) or just go with the native behavior for now? I'm OK with both.

Let's go with the native behavior for MVP.

---

Review status: 13 of 26 files reviewed at latest revision, 5 unresolved discussions.

---

*[vaadin-checkbox.html, line 57 at r10](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlUdMR6P35t0N5irhsp:-KlXRrckUisrFaWo_t8E:b-dtkds3) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/403225704c3b682d84ea7a4114a83a6bcb09f603/vaadin-checkbox.html#L57)):*
<details><summary><i>Previously, Saulis (Sauli Tähkäpää) wrote…</i></summary><blockquote>

`indeterminate` doesn't currently work, this needs to be `indeterminate="{{indeterminate::change}}"` (need to bind to a property instead of attribute + two-way binding needed). We could also add demo and tests for the indeterminate mode.
</blockquote></details>

Done, thanks!

---

*[vaadin-checkbox.html, line 119 at r10](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlUeCIGFsEpddQwf8V2:-KlXRtnUnZ9_OwpFXF3K:b-dtkds3) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/403225704c3b682d84ea7a4114a83a6bcb09f603/vaadin-checkbox.html#L119)):*
<details><summary><i>Previously, Saulis (Sauli Tähkäpää) wrote…</i></summary><blockquote>

Needs `notify: true` since `checked` changes after user action
</blockquote></details>

Done, thanks!

---

*[vaadin-checkbox.html, line 130 at r10](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlUeEDRPG5M01NnIyuj:-KlXRucESj7n-dKM3wQY:b-dtkds3) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/403225704c3b682d84ea7a4114a83a6bcb09f603/vaadin-checkbox.html#L130)):*
<details><summary><i>Previously, Saulis (Sauli Tähkäpää) wrote…</i></summary><blockquote>

Needs `notify: true` since `indeterminate` changes after user action
</blockquote></details>

Done, thanks!

---

*[vaadin-checkbox.html, line 161 at r10](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlUh1E86oeDexgXuXIM:-KlXRxRU-Xff_wvUTzSK:b-dtkds3) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/403225704c3b682d84ea7a4114a83a6bcb09f603/vaadin-checkbox.html#L161)):*
<details><summary><i>Previously, Saulis (Sauli Tähkäpää) wrote…</i></summary><blockquote>

`aria-checked` should be set to `mixed` when `indeterminate` is `true`
</blockquote></details>

Done, thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-:-KlXUAL7uEYgbEyPtSas:bbb1yby)*
<!-- Sent from Reviewable.io -->

limonte commented:
Review status: 13 of 26 files reviewed at latest revision, 5 unresolved discussions.

---

*[test/index.html, line 16 at r10](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlUfNYnm2KmsrmbPumT:-KlXWmHIBy2Fm3KqByCu:b-ww9g8c) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/403225704c3b682d84ea7a4114a83a6bcb09f603/test/index.html#L16)):*
> I think if tests don't have performance problems, and in the time that shady dom is supported we should test shady.

Agree.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1)*
<!-- Sent from Reviewable.io -->

tomivirkki commented:
*[vaadin-checkbox.html, line 60 at r11](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlXf22q2xp85CPu7BDz:-KlXf22q2xp85CPu7BE-:bj4i4wm) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/2b9f7dc7a50cf4013011261354a9f5aa502f1665/vaadin-checkbox.html#L60)):*
> ```HTML
>         tabindex="-1">
> 
>       <span part="checkbox"></span>
> ```

I wonder if this element (for custom graphical check) is needed at all. Theme can easily include custom pseudo elements and use those instead.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1)*
<!-- Sent from Reviewable.io -->

limonte commented:
Review status: 13 of 26 files reviewed at latest revision, 6 unresolved discussions.

---

*[vaadin-checkbox.html, line 60 at r11](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlXf22q2xp85CPu7BDz:-KlXgJIJQMTGqzXJaCv9:b-d35rz4) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/2b9f7dc7a50cf4013011261354a9f5aa502f1665/vaadin-checkbox.html#L60)):*
<details><summary><i>Previously, tomivirkki (Tomi Virkki) wrote…</i></summary><blockquote>

I wonder if this element (for custom graphical check) is needed at all. Theme can easily include custom pseudo elements and use those instead.
</blockquote></details>

True that, I don't have any opinion. Let's decide this later after taking a look on first versions of Material/Valo designs. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1)*
<!-- Sent from Reviewable.io -->

tomivirkki commented:
Reviewed 13 of 13 files at r11.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1)*
<!-- Sent from Reviewable.io -->

manolo commented:
Reviewed 13 of 13 files at r11.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1)*
<!-- Sent from Reviewable.io -->

Saulis commented:
Review status: 25 of 26 files reviewed at latest revision, all discussions resolved.

---

*[vaadin-checkbox.html, line 145 at r10](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlUeOj8ts06FiY_nDgU:-KlXE5pxYpambeqKCgS2:b-9r8w1b) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/403225704c3b682d84ea7a4114a83a6bcb09f603/vaadin-checkbox.html#L145)):*
<details><summary><i>Previously, manolo (Manuel Carrasco Moñino) wrote…</i></summary><blockquote>

It's needed in iron-form. So it's here for API doc purposes. Iron-form will ignore this if it does not have name+value. 
</blockquote></details>

As discussed, let's add default value `on`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1)*
<!-- Sent from Reviewable.io -->

limonte commented:
Review status: 23 of 26 files reviewed at latest revision, 1 unresolved discussion.

---

*[vaadin-checkbox.html, line 145 at r10](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-KlUeOj8ts06FiY_nDgU:-KlYT1qCuUIelrnhIbrb:b-896fix) ([raw file](https://github.com/vaadin/vaadin-checkbox/blob/403225704c3b682d84ea7a4114a83a6bcb09f603/vaadin-checkbox.html#L145)):*
<details><summary><i>Previously, Saulis (Sauli Tähkäpää) wrote…</i></summary><blockquote>

As discussed, let's add default value `on`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1)*
<!-- Sent from Reviewable.io -->

tomivirkki commented:
Reviewed 3 of 3 files at r12.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1)*
<!-- Sent from Reviewable.io -->

Saulis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 6 files at r1, 2 of 15 files at r4, 7 of 14 files at r6, 11 of 13 files at r11, 3 of 3 files at r12.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/vaadin-checkbox/1#-:-KlYUSikT9xox2jJbsZn:bnfp4nl)*
<!-- Sent from Reviewable.io -->
